### ISSUE TITLE
Add a negate flag

### ISSUE THREAD

**Issue by petemounce** [codingjoe/relint#15]
I would like to add a rule that fails when a regex is _not_ matched, for example a `VERSION` file containing a `v1.0.0` and a pattern like `^v` (because a v prefix on a semver is not valid).

**Comment by petemounce** [comment]
---
# https://semver.org/#is-v123-a-semantic-version
- name: VERSION files must contain semver.
  # LOLWAT? https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
  pattern: '^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$'
  hint: "https://semver.org/#is-v123-a-semantic-version"
  filePattern: '.*VERSION'

- name: VERSION file contents must not be prefixed with a v
  pattern: '^v'
  hint: "https://semver.org/#is-v123-a-semantic-version"
  filePattern: '.*VERSION'
```

`cmd/VERSION`:

```
v0.1.1
```

I want that file to fail both rules, not just the 2nd one.

**Comment by petemounce** [comment]
(and the LOLWAT refers to the regex itself ;) )

**Comment by codingjoe** [comment]
Hi @petemounce thanks for reaching out. I see your point, how this might come in handy. However, I am not 100% sure on how to integrate this, since we currently work with matches, that have line numbers and snippets, that are being displayed. Inverse matches would mean we only had file names.

Anyhow, I still think it's possible to come up with a good solution. How about you propose a solution. Preferably as a pull-request and I will gladly review and release it.

Best
-Joe

**Comment by viktorkertesz** [comment]
Hi @petemounce ,
did you try this? (I know it doesn't cover everything, just to get a hang of it)
```
^(?!\d+\.\d+\.\d+).*$
```
This won't match the correct versioning string but will trigger everything else.

**Comment by mroy-seedbox** [comment]
We have a similar challenge on our side. Here is the generalized idea:
- We know the way something _should_ look.
- But it is very hard to describe the way it should _not_ look (and there are possibly an infinite number of possibilities).

Simple/silly example:
- We only want to accept the string `ABC` if it is immediately followed by either `DEF` or `123`.
- The regex to find `ABC` is simple, but then it is much more challenging to say "anything else than `DEF` or `123` afterwards.

What I would propose:
- We keep the current concept of `pattern` matching to find rows in a file.
- But then we add a new concept/parameter of `mustMatch`, which would be a _positive_ regex of how the match should look, rather than trying to do it all via a _negative/exclusion_ regex.
- `pattern` would continue to match on rows, and extract the corresponding substring(s).
- `mustMatch` would then have to find an exact match (_**in each substring**_), or the rule would be triggered (for each substring).

So the solution to the above would become:
```yaml
pattern: 'ABC.*' # or maybe 'ABC.{3}'
mustMatch: 'ABC((DEF)|(123))'
```

And @petemounce, the solution for you could probably look something like this (although I'm probably missing a lot of details):
```yaml
pattern: '.\d+\.\d+\.\d+.*$' # Note the . at the beginning
mustMatch: '[^v]\d+\.\d+\.\d+'
```

So much simpler. 💙 

@petemounce, @codingjoe: What do you think?

**Comment by codingjoe** [comment]
Hm… that is an interesting idea. I like the concept of either looking for something that shouldn't be there, or should. I'd love to see this included in a new version. Heck, I'd actually love to implement path to test functions written in Python, like `test_func: "package.module:function"` and that function gets called with a file's contents.

I believe it's time for a new major release, with some exciting new features and maybe some breaking changes. If anyone is interested in contributing, please let me know.

**Comment by mroy-seedbox** [comment]
Quick idea: reLint could look for a file called `.relint.py` in the project root (just like it currently looks for `.relint.yml`), and source the python module/functions from there. So if we configure `test_func: some_function`, then it would simply call the `some_function` function from `.relint.py`. This also leaves room for a more advanced implementation via `package.module:function` in  the future.

The functions should probably have a signature along the line of `(rule, file, match)`, where:
- `rule` is the full corresponding object defined in `.relint.yml` (with `name`, `pattern`, etc.)
- `file` is the path of the file currently being linted.
  - This could be used to read the file as a string, _if necessary_.
- `match` is an individual match for the rule's `pattern` (if `pattern` is used in combination with `test_func`).
  - The function should be called once for each match (or called only once if the rule has no `pattern`, with `none` as the `match`).

Related:
- #108

**Comment by codingjoe** [comment]
Hi there 👋,

Yes, the idea of adding declarative tests has been floating around for a while. I am not opposed to the idea. However, I feature can coexist on their own. Don't hesitate to open a separate issue or discussion for such a feature.

Cheers!
Joe
